File: frontend/src/api/statsService.js
import { apiClient } from './config';

/**
 * Statistics Service
 * Handles all statistics and analytics operations
 */

/**
 * Get comprehensive dashboard statistics
 * @returns {Promise<Object>} Dashboard stats (books, users, loans)
 */
export const getDashboardStats = async () => {
  try {
    const response = await apiClient.get('/stats/dashboard');
    return response.data;
  } catch (error) {
    console.error('Get dashboard stats error:', error);
    throw error;
  }
};

/**
 * Get detailed book statistics
 * @returns {Promise<Object>} Book stats (total books, copies, most borrowed)
 */
export const getBookStats = async () => {
  try {
    const response = await apiClient.get('/stats/books');
    return response.data;
  } catch (error) {
    console.error('Get book stats error:', error);
    throw error;
  }
};

/**
 * Get most borrowed books
 * @param {number} limit - Maximum number of books to return (default 10)
 * @returns {Promise<Array>} List of most borrowed books with borrow counts
 */
export const getMostBorrowedBooks = async (limit = 10) => {
  try {
    const response = await apiClient.get('/stats/books/most-borrowed', {
      params: { limit }
    });
    return response.data;
  } catch (error) {
    console.error('Get most borrowed books error:', error);
    throw error;
  }
};

/**
 * Get detailed loan statistics
 * @param {number} year - Year for monthly loan stats (optional, defaults to current year)
 * @returns {Promise<Object>} Loan stats (active, overdue, pending, by month, top borrowers)
 */
export const getLoanStats = async (year = null) => {
  try {
    const params = year ? { year } : {};
    const response = await apiClient.get('/stats/loans', { params });
    return response.data;
  } catch (error) {
    console.error('Get loan stats error:', error);
    throw error;
  }
};

/**
 * Get loan count by month for a specific year
 * @param {number} year - Year (optional, defaults to current year)
 * @returns {Promise<Array>} Monthly loan counts
 */
export const getLoansByMonth = async (year = null) => {
  try {
    const params = year ? { year } : {};
    const response = await apiClient.get('/stats/loans/by-month', { params });
    return response.data;
  } catch (error) {
    console.error('Get loans by month error:', error);
    throw error;
  }
};

/**
 * Get users with most loans
 * @param {number} limit - Maximum number of users to return (default 10)
 * @returns {Promise<Array>} List of top borrowers with loan counts
 */
export const getTopBorrowers = async (limit = 10) => {
  try {
    const response = await apiClient.get('/stats/loans/top-borrowers', {
      params: { limit }
    });
    return response.data;
  } catch (error) {
    console.error('Get top borrowers error:', error);
    throw error;
  }
};

/**
 * Get detailed user statistics
 * @returns {Promise<Object>} User stats (total users, by role, blacklisted, with infractions)
 */
export const getUserStats = async () => {
  try {
    const response = await apiClient.get('/stats/users');
    return response.data;
  } catch (error) {
    console.error('Get user stats error:', error);
    throw error;
  }
};

/**
 * Get list of users with infractions > 0 (Admin only)
 * @returns {Promise<Array>} List of users with infractions
 */
export const getUsersWithInfractions = async () => {
  try {
    const response = await apiClient.get('/stats/users/infractions');
    return response.data;
  } catch (error) {
    console.error('Get users with infractions error:', error);
    throw error;
  }
};

export default {
  getDashboardStats,
  getBookStats,
  getMostBorrowedBooks,
  getLoanStats,
  getLoansByMonth,
  getTopBorrowers,
  getUserStats,
  getUsersWithInfractions,
};

File: backend/src/Services/statsService.py
from typing import List
from ..Brokers.statsBroker import StatsBroker


class StatsService:
    def __init__(self, broker: StatsBroker):
        self.broker = broker

    # ==================== DASHBOARD STATISTICS ====================

    async def get_dashboard_stats(self) -> dict:
        """Get comprehensive dashboard statistics"""
        return {
            "books": {
                "total_books": await self.broker.get_total_books(),
                "total_copies": await self.broker.get_total_copies(),
                "available_copies": await self.broker.get_available_copies(),
                "copies_by_status": await self.broker.get_books_by_status(),
            },
            "users": {
                "total_users": await self.broker.get_total_users(),
                "users_by_role": await self.broker.get_users_by_role(),
                "blacklisted_users": await self.broker.get_blacklisted_users_count(),
            },
            "loans": {
                "active_loans": await self.broker.get_total_active_loans(),
                "overdue_loans": await self.broker.get_total_overdue_loans(),
                "pending_requests": await self.broker.get_total_pending_requests(),
                "loans_by_status": await self.broker.get_loans_by_status(),
            },
        }

    # ==================== BOOK STATISTICS ====================

    async def get_book_stats(self) -> dict:
        """Get detailed book statistics"""
        return {
            "total_books": await self.broker.get_total_books(),
            "total_copies": await self.broker.get_total_copies(),
            "available_copies": await self.broker.get_available_copies(),
            "copies_by_status": await self.broker.get_books_by_status(),
            "most_borrowed": await self.broker.get_most_borrowed_books(limit=10),
        }

    async def get_most_borrowed_books(self, limit: int = 10) -> List[dict]:
        """Get most borrowed books"""
        return await self.broker.get_most_borrowed_books(limit)

    # ==================== LOAN STATISTICS ====================

    async def get_loan_stats(self, year: int = None) -> dict:
        """Get detailed loan statistics"""
        return {
            "active_loans": await self.broker.get_total_active_loans(),
            "overdue_loans": await self.broker.get_total_overdue_loans(),
            "pending_requests": await self.broker.get_total_pending_requests(),
            "loans_by_status": await self.broker.get_loans_by_status(),
            "loans_by_month": await self.broker.get_loans_by_month(year),
            "top_borrowers": await self.broker.get_top_borrowers(limit=10),
        }

    async def get_loans_by_month(self, year: int = None) -> List[dict]:
        """Get loan count by month"""
        return await self.broker.get_loans_by_month(year)

    async def get_top_borrowers(self, limit: int = 10) -> List[dict]:
        """Get top borrowers"""
        return await self.broker.get_top_borrowers(limit)

    # ==================== USER STATISTICS ====================

    async def get_user_stats(self) -> dict:
        """Get detailed user statistics"""
        return {
            "total_users": await self.broker.get_total_users(),
            "users_by_role": await self.broker.get_users_by_role(),
            "blacklisted_users": await self.broker.get_blacklisted_users_count(),
            "users_with_infractions": await self.broker.get_users_with_infractions(),
        }

    async def get_users_with_infractions(self) -> List[dict]:
        """Get users with infractions"""
        return await self.broker.get_users_with_infractions()
